spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -779,10 +800,54 @@ public static TableIdentifier identifierToTableIdentifier(Identifier identifier)
             Object value = CatalystTypeConverters.convertToScala(catalystValue, field.dataType());
             values.put(field.name(), value.toString());
           });
-          return new SparkPartition(values, partition.path().toString(), format);
+          FileStatus fileStatus =
+              scala.collection.JavaConverters.seqAsJavaListConverter(partition.files()).asJava().get(0);
+
+          return new SparkPartition(values, fileStatus.getPath().getParent().toString(), format);
         }).collect(Collectors.toList());
   }
 
+  private static List getPartitionFilterExpressions(StructType schema,
+                                                    Map<String, String> partitionFilter) {
+    List<org.apache.spark.sql.catalyst.expressions.Expression> filterExpressions = new java.util.ArrayList<>();
+    for (Map.Entry<String, String> entry : partitionFilter.entrySet()) {
+      try {
+        // IllegalArgumentException is thrown if schema doesn't contain this entry,
+        // which means partition filter is not on partition columns.
+        int index = schema.fieldIndex(entry.getKey());
+        org.apache.spark.sql.types.DataType dataType = schema.fields()[index].dataType();
+        BoundReference ref = new BoundReference(index, dataType, true);
+        if (dataType.sameType(DataTypes.IntegerType) || dataType.sameType(DataTypes.ShortType) ||
+            dataType.sameType(DataTypes.ByteType)) {
+          filterExpressions.add(new EqualTo(ref,
+              org.apache.spark.sql.catalyst.expressions.Literal.create(Integer.parseInt(entry.getValue()),
+              DataTypes.IntegerType)));
+        } else if (schema.fields()[index].dataType().sameType(DataTypes.StringType)) {
+          filterExpressions.add(new EqualTo(ref,
+              org.apache.spark.sql.catalyst.expressions.Literal.create(entry.getValue(), DataTypes.StringType)));
+        } else if (schema.fields()[index].dataType().sameType(DataTypes.LongType)) {
+          filterExpressions.add(new EqualTo(ref,
+              org.apache.spark.sql.catalyst.expressions.Literal.create(Long.parseLong(entry.getValue()),
+              DataTypes.LongType)));
+        } else if (schema.fields()[index].dataType().sameType(DataTypes.DateType)) {
+          filterExpressions.add(new EqualTo(ref,
+              org.apache.spark.sql.catalyst.expressions.Literal.create(entry.getValue(), DataTypes.DateType)));
+        } else if (schema.fields()[index].dataType().sameType(DataTypes.TimestampType)) {
+          filterExpressions.add(new EqualTo(ref,
+              org.apache.spark.sql.catalyst.expressions.Literal.create(entry.getValue(), DataTypes.TimestampType)));
+        } else if (schema.fields()[index].dataType()
+            .sameType(DataTypes.CalendarIntervalType)) {
+          filterExpressions.add(new EqualTo(ref,
+              org.apache.spark.sql.catalyst.expressions.Literal.create(entry.getValue(),
+              DataTypes.CalendarIntervalType)));
```

Comment:
I only tested dataType Integer and String by using the existing test suite `TestAddFilesProcedure`. If this PR looks OK, I will also test all the other data types.

Reply:
For this large if-else-if chain, you might want to look into this lookup-map pattern used here: https://github.com/apache/iceberg/blob/466073b7d8c23ebeae045822ee6e1a1104a5ed5a/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkFilters.java#L80-L220

I'm not sure if a look-up map can be used here because of the usage of `sameType` function, but it might be worth looking into 😄